tfhe/src/integer/server_key/radix_parallel/add.rs

```diff
@@ -386,188 +398,956 @@ impl ServerKey {
+                                let output_carry =
+                                    ((lhs + rhs + input_carry) >> num_bits_in_message) & 1;
+
+                                let input_carry_to_last_bit =
+                                    ((lhs_except_last_bit + rhs_except_last_bit + input_carry)
+                                        >> (num_bits_in_message - 1))
+                                        & 1;
+
+                                u64::from(input_carry_to_last_bit != output_carry)
+                            };
+
+                            (overflows_with_given_input_carry(1) << 3)
+                                | (overflows_with_given_input_carry(0) << 2)
+                        });
+                        self.key.apply_lookup_table_assign(overflow_flag, &lut);
+                    });
+                }
             });
-        assert!(degree_after_add_does_not_go_beyond_first_carry);
 
-        self.unchecked_add_assign_parallelized(lhs, rhs);
-        self.propagate_single_carry_parallelized_low_latency(lhs.blocks_mut())
+            let num_blocks = lhs.len();
+
+            // We did the first block before, the last block is done after this if,
+            // so we need 3 blocks at least to enter this
+            if num_blocks >= 3 {
+                for (lhs_b, rhs_b) in lhs[1..num_blocks - 1]
+                    .iter_mut()
+                    .zip(rhs[1..num_blocks - 1].iter())
+                {
+                    self.key.unchecked_add_assign(lhs_b, rhs_b);
+                    self.key.unchecked_add_assign(lhs_b, &carry);
+
+                    carry.clone_from(lhs_b);
+                    rayon::join(
+                        || self.key.message_extract_assign(lhs_b),
+                        || self.key.carry_extract_assign(&mut carry),
+                    );
+                }
+            }
+
+            if num_blocks >= 2 {
+                // Handle the last block
+                self.key
+                    .unchecked_add_assign(&mut lhs[num_blocks - 1], &rhs[num_blocks - 1]);
+                self.key
+                    .unchecked_add_assign(&mut lhs[num_blocks - 1], &carry);
+            }
+
+            if let Some(block) = overflow_flag.as_mut() {
+                if num_blocks == 1 && input_carry.is_some() {
+                    self.key
+                        .unchecked_add_assign(block, input_carry.map(|b| &b.0).unwrap());
+                } else {
+                    self.key.unchecked_add_assign(block, &carry);
+                }
+            }
+
+            // To be able to use carry_extract_assign in it
+            carry.clone_from(&lhs[num_blocks - 1]);
+
+            // Note that here when num_blocks == 1 && requested_flag != Overflow nothing
+            // will actually be spawned.
+            rayon::scope(|s| {
+                if num_blocks >= 2 {
+                    // These would already have been done when the first block was processed
```

Comment:
if num blocks == 1, then we skip this message/carry_extract, because they would have been done earlier in the function

It happens because the computation of the overflow flag requires 2 steps and to we insert these two steps in parallel to normal carry propagation. The first step is done when processing the first block, the second step is done when processing the last block. So if the number of block is smaller than 2 then the overflow computation add additional layer of PBS